ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -0,0 +1,164 @@
+-- layer to run a given client \/ server peers.
+--
+data MuxPeer m where
+  -- |
+  -- A non pipeliend peer together with a codec.
+  --
+  OnlyClient
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsClient st m a
+    -> MuxPeer m
+
+  -- |
+  -- A pipelined peer together with a codec.
+  --
+  OnlyPipelinedClient
+    :: forall ps (st :: ps) m a.
+       Natural
+    -> Codec ps CBOR.DeserialiseFailure m ByteString
+    -> PeerPipelined ps AsClient st m a
+    -> MuxPeer m
+
+  -- |
+  -- Server peer with a codec
+  OnlyServer
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+  -- |
+  -- Client and server peers with the corresponding codec.
+  --
+  ClientAndServer
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsClient st m a
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+  -- |
+  -- Pipelined client and a server with the correspnding codec.
+  --
+  PipelinedClientAndServer
+    :: forall ps (st :: ps) m a.
+       Natural
+    -> Codec ps CBOR.DeserialiseFailure m ByteString
+    -> PeerPipelined ps AsClient st m a
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+
+-- |
+-- Public network interface for 'ouroboros-network'.
+--
+data NetworkInterface ptcl addr m = NetworkInterface {
+      -- |
+      -- Address of the node to run.  The node will bind to this address, and
+      -- listen for incoming connections.  Some bearers do not have a notion of
+      -- address.
+      --
+      nodeAddress :: addr, 
+
+      -- |
+      -- Versioned set of peers to run; it is the responsibilty of the caller
+      -- to provide the correct associtation between `ptcl` and a protocol
+      -- (which is existentially quantified using @'PeerSpecification'@).
+      --
+      protocols   :: [(Version, ptcl -> MuxPeer m)]
+    }
+
+-- | Low level network interface.  It can be intiatiated using a socket, pair
+-- of pipes or a pair queues.
+--
+data NetworkNode addr m = NetworkNode {
+      -- |
+      -- The way to connect ot other peers.  On startup the network interface
+      -- will run this to connect with a given list of peer.  But it can also
+      -- be used at a later stage to connect to a new peer.
+      --
+      -- This function will run client side of mux version negotation and then
+      -- start a the list protocols given by @'NetworkInterface'@.
+      --
+      connectTo :: addr -> m (),
+
+      -- |
+      -- This will cancel the thread that is listening for new connections and
+      -- close the underlaying bearer.
+      killNode  :: m ()
+    }
+
+-- |
+-- Transform a @'MuxPeer'@ into @'ProtocolDescription'@ used by the
+-- multiplexing layer.
+--
+miniProtocolDescription
+  :: forall m ptcl.
+     ( MonadAsync m
+     , MonadCatch m
+     , MonadThrow m
+     )
+  => MuxPeer m
+  -> MiniProtocolDescription ptcl m
```

Comment:
Good. Nicer than having to supply the result of `runPeer` for the mini protocol description.